server/irmaserver/helpers.go

```diff
@@ -429,3 +504,51 @@ func (s *Server) sessionMiddleware(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r.WithContext(context.WithValue(ctx, "session", session)))
 	})
 }
+
+func (s *Server) authenticationMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		session := r.Context().Value("session").(*session)
+
+		authorization := r.Header.Get(irma.AuthorizationHeader)
+		if authorization == "" && session.version != nil {
+			// Protocol versions below 2.7 use legacy authentication handled in handleSessionGet and cacheMiddleware.
+			if session.version.Below(2, 7) {
+				next.ServeHTTP(w, r)
+			} else {
+				server.WriteError(w, server.ErrorClientUnauthorized, "No authorization header provided")
+			}
+			return
+		}
+
+		// If the client connects for the first time, we grant access and make sure
+		// only that exact client can connect to this session in future requests.
+		if session.clientAuth == "" {
+			session.clientAuth = authorization
```

Comment:
I moved this check into the `bindingMiddleware`, since it is actually the legacy check that a phone is properly bound and connected. All endpoints exept `/` have to pass the `bindingMiddleware`, so the check is actually still there. This means the code for this check is not duplicate anymore now. This means the behaviour when calling the one of the other endpoints first is exactly the same as now, it gives a 403 UNEXPECTED REQUEST. 

In this way you can say that the endpoints that only pass the `authenticationMiddleware` are endpoints that can be done as a first call (which is now only the `/` endpoint). All endpoints that also pass the `bindingMiddleware` can only be accessed after the client is fully bound and connected, the right authorization token is supplied and the session is not finished yet. This makes that it works the same as it did.

Reply:
I can of course move the check of the authorization token to the `bindingMiddleware` too, remove the `authenticationMiddleware` and put the initial check in the `/` handler.